Add a frontend interface for committing RoutineLoadTask

EmmyMiao87 opened the pull request:
1. add a needSchedulerTasksQueue in LoadManager: the RoutineLoadTaskScheduler will poll task from this queue and schedule task.
2. add a frontend interface named rlTaskCommit: commit txn, update offset and renew a task for the same partitions
3. add a extra property in transaction state: in rlTaskCommit,the extra property which looks like {"job_id": xxx, "progress": xxx}
When fe initialize routine load job meta from logs, all of txn state which related to routine load job will be used for initializing progress of job.

imay commented on gensrc/thrift/FrontendService.thrift:
```diff
@@ -531,6 +531,22 @@ struct TLoadTxnRollbackResult {
+    3: required i64 backendId
+    4: required TLoadTxnCommitRequest loadTxnCommitRequest
+    // kafka progress such as {"partitionIdToOffset":{"1":100,"2",200}}
+    5: required string progress
```
you should define a struct for Kafka load progress

EmmyMiao87 replied:
Yes, there is a kafka progress, but not here.

imay commented on gensrc/thrift/FrontendService.thrift:
```diff
@@ -531,6 +531,22 @@ struct TLoadTxnRollbackResult {
+struct TRLTaskCommitRequest {
+    1: required string taskId
+    2: required i64 signature
+    3: required i64 backendId
```
why backendId is **required**?

EmmyMiao87 replied:
I need to remove task for `AgentTaskQueue`. And the params of removeTask need backendId

imay commented on gensrc/thrift/FrontendService.thrift:
```diff
@@ -531,6 +531,22 @@ struct TLoadTxnRollbackResult {
     1: required Status.TStatus status
 }
 
+struct TRLTaskCommitRequest {
```
I think you should put this request to `TLoadTxnCommitRequest` as an optional field. Then you needn't define a new RPC interface. In addition, if you add new type of RLTask, you just add new option field.

EmmyMiao87 replied:
For me, all of type of RLTask will commit task by `This method`.  The new option field is in `TRLTaskCommitRequest` not in `TLoadTxnCommitRequest`. Txn just is  a property in RLTask instead of RLTask is a property in Txn.

imay replied:
We only commit transaction, task can be finished through callback after transaction commit process.

imay commented on fe/src/main/java/org/apache/doris/load/TransactionCallBack.java:
```diff
@@ -0,0 +1,39 @@
+     *
+     * @param txnState
+     */
+    void updateCatalogAfterCommitted(TransactionState txnState);
```
Usually, callback function's name is like `onCommit`, `onAbort`

imay commented on fe/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java:
```diff
@@ -766,6 +781,16 @@ private boolean checkTxnHasRelatedJob(TransactionState txnState, Map<Long, Set<L
             return true;
         }
 
+        if (txnState.getSourceType() == LoadJobSourceType.ROUTINE_LOAD_TASK) {
```
It's not good to judge type here. You can put a callback when you begin a transaction, and then when transaction state is changed, your callback function will be called

EmmyMiao87 commented on gensrc/thrift/FrontendService.thrift:
```diff
@@ -510,6 +510,7 @@ struct TLoadTxnCommitRequest {
     7: required i64 txnId
     8: required bool sync
     9: optional list<Types.TTabletCommitInfo> commitInfos
+    10: optional string extra
```
OK, I will change RoutineLoadTaskTxnExtra to a thrift struct

EmmyMiao87 replied:
But if I use a thrift struct which routine load task need, this is not useful for other transaction

EmmyMiao87 commented on fe/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java:
```diff
@@ -356,10 +357,23 @@ public void commitTransaction(long dbId, long transactionId, List<TabletCommitIn
+                Catalog.getCurrentCatalog().getRoutineLoadInstance().getJobByTaskId(transactionState.getLabel())
+                        .updateCatalogAfterCommitted(transactionState);
+            } catch (MetaNotFoundException e) {
+                LOG.error("failed to update routine load task {} after committed txn {}",
```
Yes, it is. If update routine load task failed after committed txn, it means that offset in routine load is in back of data in palo

EmmyMiao87 commented on fe/src/main/java/org/apache/doris/load/routineload/RoutineLoadJob.java:
```diff
@@ -241,4 +303,99 @@ private void checkStateTransform(RoutineLoadJob.JobState currentState, RoutineLo
+            LabelAlreadyExistsException, BeginTransactionException;
+
+    @Override
+    public boolean checkTxnHasRelatedJob(TransactionState txnState) {
```
This check works for broker load job also. If I change to task , I think it is confused for broker load job

EmmyMiao87 commented on fe/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java:
```diff
@@ -766,6 +781,16 @@ private boolean checkTxnHasRelatedJob(TransactionState txnState, Map<Long, Set<L
+                return Catalog.getCurrentCatalog().getRoutineLoadInstance().getJobByTaskId(txnState.getLabel())
+                        .checkTxnHasRelatedJob(txnState);
+            } catch (MetaNotFoundException e) {
+                // could not find job by task id while task has been deleted
```
I will change it

EmmyMiao87 commented on fe/src/main/java/org/apache/doris/load/routineload/RoutineLoadJob.java:
```diff
@@ -241,4 +303,99 @@ private void checkStateTransform(RoutineLoadJob.JobState currentState, RoutineLo
 
+    private void loadTxnCommit(TLoadTxnCommitRequest request) throws TException {
+        FrontendServiceImpl frontendService = new FrontendServiceImpl(ExecuteEnv.getInstance());
+        frontendService.loadTxnCommit(request);
```
Already changed

imay commented on fe/src/main/java/org/apache/doris/transaction/TransactionState.java:
```diff
@@ -188,6 +197,8 @@ public void write(DataOutput out) throws IOException {
             out.writeLong(errorReplciaId);
         }
+        // TODO(ml): extra write
+        Text.writeString(out, extra.toString());
```
Don't use toString to serialize. If you want to serialize it, you should add `write` function to `TxnStateExtra`

imay commented on fe/src/main/java/org/apache/doris/transaction/TransactionState.java:
```diff
@@ -334,8 +355,28 @@ public LoadJobSourceType getSourceType() {
+        return callBack;
+    }
+
+    private TransactionStateExtra readTxnStateExtra(DataInput in) throws IOException {
```
This should to be a static function in `TransactionStateExtra`

EmmyMiao87 commented on gensrc/thrift/FrontendService.thrift:
```diff
@@ -510,6 +510,35 @@ struct TLoadTxnCommitRequest {
     7: required i64 txnId
     8: required bool sync
     9: optional list<Types.TTabletCommitInfo> commitInfos
+    10: optional TTxnStateExtra txnStateExtra
```
Java can compile successfully but c++ cannot. I will change the order

imay commented on fe/src/main/java/org/apache/doris/transaction/TransactionState.java:
```diff
@@ -188,6 +199,15 @@ public void write(DataOutput out) throws IOException {
         for (long errorReplciaId : errorReplicas) {
             out.writeLong(errorReplciaId);
         }
+        if (Catalog.getCurrentCatalogJournalVersion() >= FeMetaVersion.VERSION_46) {
```
em...now, this is work in progress, you'd better not to persist any object. Because code would be changed later. If you persist something, there will be more work to do compatible.

imay commented on fe/src/main/java/org/apache/doris/analysis/InsertStmt.java:
```diff
@@ -236,11 +236,11 @@ public void analyze(Analyzer analyzer) throws UserException {
             transactionId = Catalog.getCurrentGlobalTransactionMgr().beginTransaction(db.getId(),
                     jobLabel,
-                    "fe", sourceType);
+                    "fe", sourceType, null);
```
Like this, you can add a `beginTransaction` with four parameters, and old version function call new version add a `null` argument.

imay commented on fe/src/main/java/org/apache/doris/load/TransactionCallBack.java:
```diff
@@ -0,0 +1,39 @@
+     *
+     * @param txnState
+     */
+    void updateCatalogOnCommitted(TransactionState txnState);
```
rename it to `onCommitted`, maybe some callback just want to print a notice other than update catalog

imay commented on fe/src/main/java/org/apache/doris/load/TransactionCallBack.java:
```diff
@@ -0,0 +1,39 @@
+
+import org.apache.doris.transaction.TransactionState;
+
+public interface TransactionCallBack {
```
```suggestion
public interface TransactionCallback {
```

imay commented on fe/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java:
```diff
@@ -766,6 +781,10 @@ private boolean checkTxnHasRelatedJob(TransactionState txnState, Map<Long, Set<L
         }
 
+        if (txnState.getSourceType() == LoadJobSourceType.ROUTINE_LOAD_TASK) {
+            return txnState.getCallBack().checkTxnHasRelatedJob(txnState);
```
Do you need to callback for check relate job?

EmmyMiao87 replied:
Yes, it need. The reason is that task will be add into a needSchedulerQueue when it is created. Before task scheduler to be, the txn  state of task  is PREPARE. At the same time,  GlobelTransactionMgr clean old transactions include PREPARE. The task and txn of task live and die together

imay commented on gensrc/thrift/FrontendService.thrift:
```diff
@@ -510,6 +538,7 @@ struct TLoadTxnCommitRequest {
     7: required i64 txnId
     8: required bool sync
     9: optional list<Types.TTabletCommitInfo> commitInfos
+    10: optional TTxnStateExtra txnStateExtra
```
This struct is only used when transaction commit, is `TTransactionCommitAttachment` better?

imay commented on fe/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java:
```diff
@@ -121,14 +127,17 @@ public long beginTransaction(long dbId, String label, String coordinator, LoadJo
             long tid = idGenerator.getNextTransactionId();
             LOG.debug("beginTransaction: tid {} with label {} from coordinator {}", tid, label, coordinator);
             TransactionState transactionState = new TransactionState(dbId, tid, label, sourceType, coordinator);
+            if (transactionCallback != null) {
```
I think you can add callback to `TransactionState` constructor to avoid this `if` statement. And I prefer `TransactionStateChangeListener` to `Callback`

EmmyMiao87 replied:
I will change it.

imay commented on fe/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java:
```diff
@@ -186,11 +195,11 @@ public void deleteTransaction(long transactionId) {
      */
     public void commitTransaction(long dbId, long transactionId, List<TabletCommitInfo> tabletCommitInfos)
-            throws MetaNotFoundException, TransactionCommitFailedException {
+            throws MetaNotFoundException, TransactionException {
```
I think you should add a parameter for `CommitAttachment`

imay commented on fe/src/main/java/org/apache/doris/common/FeMetaVersion.java:
```diff
@@ -97,4 +97,7 @@
     public static final int VERSION_45 = 45;
+
+    // add extra in transaction state for routine load
+    public static final int VERSION_46 = 46;
```
remove this VERSION_46, you can you VERSION_RL in commented code instead

morningman commented on fe/src/main/java/org/apache/doris/load/routineload/RoutineLoadTaskInfo.java:
```diff
@@ -53,4 +68,16 @@ public void setLoadStartTimeMs(long loadStartTimeMs) {
+    abstract RoutineLoadTask createStreamLoadTask(long beId);
+
+    @Override
+    public boolean equals(Object obj) {
```
What if obj is not an instance of  RoutineLoadTaskInfo?
You should check it first

morningman commented on fe/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java:
```diff
@@ -448,6 +474,8 @@ public void abortTransaction(Long dbId, String label, String reason) throws User
                 }
             }
+        } catch (TransactionException e) {
+            LOG.warn("failed to update transaction state to aborted");
```
This kind of log is meaningless, no one knows which transaction is failed.

morningman commented on fe/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java:
```diff
@@ -527,6 +555,8 @@ public void finishTransaction(long transactionId, Set<Long> errorReplicaIds) {
                 unprotectUpsertTransactionState(transactionState);
                 return;
+            } catch (TransactionException e) {
+                LOG.warn("failed to change transaction status to aborted");
```
This kind of log is meaningless, no one knows which transaction is aborted.

morningman commented on fe/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java:
```diff
@@ -646,6 +676,8 @@ public void finishTransaction(long transactionId, Set<Long> errorReplicaIds) {
                 transactionState.setTransactionStatus(TransactionStatus.VISIBLE);
                 unprotectUpsertTransactionState(transactionState);
+            } catch (TransactionException e) {
+                LOG.warn("failed to change txn state to visible");
```
This kind of log is meaningless, no one knows which transaction is visible.

kangpinghuang commented on fe/src/main/java/org/apache/doris/analysis/InsertStmt.java:
```diff
@@ -137,93 +137,93 @@ public long getTransactionId() {
-
-
+    
+    
```
additional empty line

kangpinghuang commented on fe/src/main/java/org/apache/doris/analysis/InsertStmt.java:
```diff
@@ -59,11 +59,11 @@
 public class InsertStmt extends DdlStmt {
     private static final Logger LOG = LogManager.getLogger(InsertStmt.class);
-
+    
```
additional space?

kangpinghuang commented on fe/src/main/java/org/apache/doris/analysis/RoutineLoadStmt.java:
```diff
@@ -0,0 +1,49 @@
+*/
+
+public class RoutineLoadStmt {
+}
```
no class body?

imay reviewed (dismissed):
lgtm

EmmyMiao87 committed:
Add a frontend interface for committing RoutineLoadTask

1. add a needSchedulerTasksQueue in LoadManager: the RoutineLoadTaskScheduler will poll task from this queue and schedule task.
2. add a frontend interface named rlTaskCommit: commit txn, update offset and renew a task for the same partitions
3. add extra property in transaction state: in rlTaskCommit, extra property which looks like {"job_id": xxx, "progress": xxx}
When fe initialize routine load job meta from logs, all of txn state which related to routine load job will be used for initializing progress of job.

Add a TxnStateChangeListener interface for transaction
1. onCommitted , onAborted, beforeAborted will be called by different type of txn
2. RoutineLoadJob will update job progress and renew a task when onCommitted callback
3. Add TxnStateChangeListener into TransactionState
4. set transactionState to committed will call onCommitted callback if callback is not null
5. set transactionState to aborted will call beforeAborted and onAborted
6. beforeAborted in RoutineLoadJob will check if there is related task when TxnStatusChangeReason is TIMEOUT. It will prevent abort when there is a related task by throw TransactionException
7. Other reason of abort will not prevent abort. The onAborted will be call and job state will be change to paused

Change extra to TxnCommitAttachment in TLoadTxnCommitRequest
1. The KAFKA source of TTxnSourceType means that this is a routine load task commit. And the TRLTaskTxnCommitAttachment is the commitInfo of this task.
2. TRLTaskTxnCommitAttachment will be convert to RLTaskTxnCommitAttachment which include progress of this task, task id, numOfErrorData etc.

Add param TxnCommitAttachment into commitTransaction
1. The TxnCommitAttachment will be updated in commitTransaction
